tools/sdk-generation-pipeline/packages/sdk-generation-lib/src/lib/ArtifactUploader.ts

```diff
@@ -0,0 +1,189 @@
+                        await this.azureBlobClient.publishBlob(
+                            path.join(packageFolder, filePath),
+                            `${this.pipelineBuildId}/${this.language}/${packageName}/${filePath}`
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    public async uploadArtifacts(generateAndBuildOutputJson: GenerateAndBuildOutput) {
+        for (const p of generateAndBuildOutputJson.packages) {
+            const result = p.result;
+            if (result === 'failed') {
+                logger.warn(`Build ${p.packageName} failed, skipped it`);
+                continue;
+            }
+            const artifacts = p.artifacts;
+            if (!artifacts) {
+                // artifacts is optional
+                continue;
+            }
+
+            for (const artifact of artifacts) {
+                const artifactName = path.basename(artifact);
+                await this.azureBlobClient.publishBlob(
+                    artifact,
+                    `${this.pipelineBuildId}/${this.language}/${artifactName}`
+                );
+            }
+        }
+    }
+}
+
+export type ArtifactPipelineUploaderContext = {
+    artifactDir: string;
+    language: string;
+};
+
+export class ArtifactPipelineUploader {
+    private artifactDir: string;
+    private language: string;
+
+    constructor(artifactPipelineUploaderContext: ArtifactPipelineUploaderContext) {
+        this.validateContext(artifactPipelineUploaderContext);
+        this.artifactDir = artifactPipelineUploaderContext.artifactDir;
+        this.language = artifactPipelineUploaderContext.language;
+    }
+
+    private validateContext(artifactPipelineUploaderContext: ArtifactPipelineUploaderContext) {
+        if (
+            !artifactPipelineUploaderContext.artifactDir ||
+            !fs.existsSync(artifactPipelineUploaderContext.artifactDir)
+        ) {
+            throw new Error(`Invalid artifactDir`);
+        }
+        if (!(<any>Object).values(SDK).includes(artifactPipelineUploaderContext.language)) {
+            throw new Error(`Invalid language`);
+        }
+    }
+
+    public async publishSourceCode(generateAndBuildOutputJson: GenerateAndBuildOutput) {
+        for (const p of generateAndBuildOutputJson.packages) {
+            const result = p.result;
+            if (result === 'failed') {
+                logger.warn(`Build ${p.packageName} failed, skipped it`);
+                continue;
+            }
+            const packageName = p.packageName;
+            const packageFolder = p.packageFolder;
+
+            if (packageFolder && fs.existsSync(packageFolder)) {
+                for (const filePath of getFileListInPackageFolder(packageFolder)) {
+                    if (fs.existsSync(path.join(packageFolder, filePath))) {
+                        const fileDir = path.dirname(filePath);
+                        child_process.execSync(
```

Comment:
As a general principle we should think twice when we are shelling out to child process in our tools/utils. If we find we are doing this we should probably instead be doing this task in a shell script (bash or PS).

Reply:
yes, the purpose of this function is copying files to the DevOps Publish pipeline artifacts output directory. I will replace the child process with `fs.mkdirSync` and `fs.copyFileSync`.